### PR TITLE
bluetooth: controller: Revised ticker for improved conflict resolution

### PR DESCRIPTION
**Pull request by mtpr-ot**
Changed ticker behavior to resolve conflicts in ticker_worker instead of
in ticker_job. This allows better real-time slot allocation, as well as
allowing callbacks even if ticker node doesn't get air-time (for audio
stream housekeeping). A priority property now also allows prioritizing
one ticker node over others.
Compatibility mode enables lagacy ticker behavior if needed.

Signed-off-by: Morten Priess <mtpr@oticon.com>

### CODE REVIEW

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 1]
```diff
@@ -292,7 +309,7 @@ static void ticker_by_next_slot_get(struct ticker_instance *instance,
 	u32_t _ticks_to_expire;
 	u8_t _ticker_id_head;
 
-	node = instance->nodes;
+	node = instance->node;
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ticker/ticker.c]: revert these changes, i think these changes where approve upstream

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 2]
```diff
@@ -536,10 +747,24 @@ void ticker_worker(void *param)
 	ticker_id_head = instance->ticker_id_head;
 
 	/* Expire all tickers within ticks_elapsed and collect ticks_expired */
-	node = &instance->nodes[0];
+	node = &instance->node[0];
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ticker/ticker.c]: move this close to the while loop

**Review by cvinayak — COMMENTED** [review]
first glance, it looks good, I need time until 19/7 to do some testing.

**Comment by mtpr-ot** [comment]
@cvinayak, any idea why Shippable fails?

**Comment by aescolar** [comment]
CC @alexeRadu

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 3]
```diff
@@ -535,11 +746,25 @@ void ticker_worker(void *param)
 	/* Auto variable containing the head of tickers expiring */
 	ticker_id_head = instance->ticker_id_head;
 
+	/* Check if the previous ticker node which had air-time, is still
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ticker/ticker.c]: Missing #if !defined(CONFIG_BT_TICKER_COMPATIBILITY_MODE)

**Review by cvinayak — CHANGES_REQUESTED** [review]

**Comment by cvinayak** [comment]
The changes look good by inspection (minor change requested) and sanity run of scan_adv sample on nRF51 seems ok.

File: subsys/bluetooth/controller/ticker/ticker.c (hunk 3 above)
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ticker/ticker.c]: Will add the #if to avoid warnings. Is it possible for you to test scan_adv with nRF52 as well (=> new code exercised). As far as I can see, advertising is identical to legacy ticker.

**Comment by mtpr-ot** [comment]
> The changes look good by inspection (minor change requested) and sanity run of scan_adv sample on nRF51 seems ok.

. Is it possible for you to test scan_adv with nRF52 as well (=> new code exercised)? As far as I can see with nRF52, advertising is identical to using legacy ticker.

**Comment by mtpr-ot** [comment]
@cvinayak Ready!

**Review by cvinayak — APPROVED** [review]
LGTM

**Review by aescolar — APPROVED** [review]

**Comment by aescolar** [comment]
As discussed on Tuesday 2019/08/06, this PR should correct the issue found in the previous PR #15787 (which was followed by the revert in #16885), by disabling the new functionality by default for the NRF51 with the new BT_TICKER_COMPATIBILITY_MODE.
If we were to find issues for NRF52 platforms, we could (in the worst case) change that default also for those.
